File: backend/src/core/ai/services/pinecone.service.ts
import { Injectable } from '@nestjs/common';
import { Pinecone } from '@pinecone-database/pinecone';
import { OpenAIEmbeddings } from '@langchain/openai';

@Injectable()
export class PineconeService {
    private pinecone: Pinecone;
    private embedding: OpenAIEmbeddings;
  
    constructor() {
        this.pinecone = new Pinecone({ apiKey: process.env.PINECONE_API_KEY });
        this.embedding = new OpenAIEmbeddings({
          modelName: "text-embedding-ada-002",
          openAIApiKey: process.env.OPENAI_API_KEY,
        });

    }

    async generateEmbedding(textsToEmbed: string[]) {
        return await this.embedding.embedDocuments(textsToEmbed);
    }

    async createVectors(vectors) {
        const index = this.pinecone.Index("exclusive-index");
        await index.upsert(vectors);
    }

}

File: ai-microservice/app/pinecone_service.py
from langchain_openai import OpenAIEmbeddings
from pinecone import Pinecone
from app import config

class PineconeService:
    def __init__(self):
        self.pinecone = Pinecone(api_key=config.PINECONE_API_KEY)
        self.embedding = OpenAIEmbeddings(
            model="text-embedding-ada-002",
            openai_api_key=config.OPENAI_API_KEY
        )

    def get_embedding(self, query: str):
        return self.embedding.embed_query(query)

    def query_index(self, vector):
        index = self.pinecone.Index("exclusive-index")
        results = index.query(
            vector=vector,
            top_k=3,
            include_metadata=True
        )
        return results
